src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxPermissionFilter.cs

```diff
@@ -0,0 +1,118 @@
+                var txPermissions = GetPermissions(tx, blockHeader);
+                if (_logger.IsTrace) _logger.Trace($"Given transaction data: sender: {tx.SenderAddress} to: {tx.To} value: {tx.Value}, gas_price: {tx.GasPrice}. Permissions required: {txType}, got: {txPermissions}.");
+                return (txPermissions & txType) == txType;
+            }
+        }
+
+        private TransactionPermissionContract.TxPermissions GetPermissions(Transaction tx, BlockHeader blockHeader)
+        {
+            var key = (blockHeader.Hash, tx.SenderAddress);
+            var txCachedPermissions = _cache.Permissions.Get(key);
+            return txCachedPermissions ?? GetPermissionsFromContract(tx, blockHeader, key);
+        }
+
+        private TransactionPermissionContract.TxPermissions GetPermissionsFromContract(Transaction tx, BlockHeader blockHeader, in (Keccak Hash, Address SenderAddress) key)
+        {
+            TransactionPermissionContract.TxPermissions txPermissions = TransactionPermissionContract.TxPermissions.None;
+            bool cache = true;
+            
+            var versionedContract = GetVersionedContract(blockHeader);
+            if (versionedContract == null)
+            {
+                if (_logger.IsError) _logger.Error("Unknown version of tx permissions contract is used.");
+            }
+            else
+            {
+                if (_logger.IsTrace) _logger.Trace($"Version of tx permission contract: {versionedContract.Version}.");
+                
+                try
+                {
+                    (txPermissions, cache) = versionedContract.AllowedTxTypes(blockHeader, tx);
+                }
+                catch (AuRaException e)
+                {
+                    if (_logger.IsError) _logger.Error("Error calling tx permissions contract.", e);
+                }
+            }
+
+            if (cache)
+            {
+                _cache.Permissions.Set(key, txPermissions);
+            }
+
+            return txPermissions;
+        }
+
+        private TransactionPermissionContract.ITransactionPermissionVersionedContract GetVersionedContract(BlockHeader blockHeader)
+        {
+            TransactionPermissionContract.ITransactionPermissionVersionedContract versionedContract;
+            if (_cache.VersionedContracts.TryGet(blockHeader.Hash, out var version))
+            {
+                versionedContract = _contract.GetVersionedContract(version);
+            }
+            else
+            {
+                versionedContract = _contract.GetVersionedContract(blockHeader);
+                _cache.VersionedContracts.Set(blockHeader.Hash, versionedContract.Version);
+            }
+
+            return versionedContract;
+        }
+
+        private TransactionPermissionContract.TxPermissions GetTxType(Transaction tx) =>
+            tx.IsContractCreation 
+                ? TransactionPermissionContract.TxPermissions.Create 
+                : tx.IsMessageCall 
```

Comment:
PROBABLY MOST IMPORTANT HERE: how do you recognize if it is  message call or basic - I think the current IsMessageCall is broken

Reply:
Fixed, we need to check state if its contract receipient